File: webapp/src/Websocket/types/Subscription.ts
import { UnderFloorHeaterResponse } from "./UnderFloorHeater";
import { EvChargingEvent } from "./EVChargingStation";
import { EnvironmentSensorEvent } from "./EnvironmentSensors";
import { QuickStatsResponse } from "./QuickStats";
import { ESSState } from "./EnergyStorageSystem";
import {FirewallState, DnsBlockingState} from "./Firewall";
import {GarageLightStatus} from "./Garage";
import {HoermannE4Broadcast} from "./garage.domain";

export type SubscriptionType =
    "getGarageLightStatus"
    | "getGarageDoorStatus"
    | "getUnderFloorHeaterStatus"
    | "evChargingStationEvents"
    | "environmentSensorEvents"
    | "quickStatsEvents"
    | "firewall"
    | "essState";

export type Subscribe = {
    subscriptionId: string;
    type: SubscriptionType;
}

export type Unsubscribe = {
    subscriptionId: string;
}

export type Notify = {
    subscriptionId: string;
    garageStatus?: GarageLightStatus;
    underFloorHeaterStatus?: UnderFloorHeaterResponse;
    evChargingStationEvent?: EvChargingEvent;
    environmentSensorEvent?: EnvironmentSensorEvent;
    quickStatsResponse?: QuickStatsResponse;
    essState?: ESSState;
    firewallState?: FirewallState;
    hoermannE4Broadcast?: HoermannE4Broadcast;
}

File: src/main/kotlin/com/dehnes/smarthome/api/dtos/Subscription.kt
package com.dehnes.smarthome.api.dtos

import com.dehnes.smarthome.firewall_router.FirewallState
import com.dehnes.smarthome.garage.GarageVentilationState
import com.dehnes.smarthome.garage.HoermannE4Broadcast
import com.dehnes.smarthome.victron.ESSState

enum class SubscriptionType {
    getGarageLightStatus,
    getGarageDoorStatus,
    getGarageVentilationStatus,
    getUnderFloorHeaterStatus,
    evChargingStationEvents,
    environmentSensorEvents,
    quickStatsEvents,
    essState,
    firewall
}

data class Subscribe(
    val subscriptionId: String,
    val type: SubscriptionType,
)

data class Unsubscribe(
    val subscriptionId: String
)

data class Notify(
    val subscriptionId: String,
    val garageStatus: GarageLightStatus? = null,
    val underFloorHeaterStatus: UnderFloorHeaterResponse? = null,
    val evChargingStationEvent: EvChargingEvent? = null,
    val environmentSensorEvent: EnvironmentSensorEvent? = null,
    val quickStatsResponse: QuickStatsResponse? = null,
    val essState: ESSState? = null,
    val firewallState: FirewallState? = null,
    val hoermannE4Broadcast: HoermannE4Broadcast? = null,
    val garageVentilationState: GarageVentilationState? = null,
)
